include/swift/SILOptimizer/Analysis/ARCAnalysis.h

```diff
@@ -386,4 +390,179 @@ SILInstruction *findReleaseToMatchUnsafeGuaranteedValue(
+class EpilogueARCContext {
+public:
+  enum EpilogueARCKind { Retain = 0, Release = 1 };
+
+private:
+  // Are we finding retains or releases.
+  EpilogueARCKind Kind;
+
+  // The argument we are looking for epilogue ARC instruction for.
+  SILValue Arg;
+
+  /// The allocator we are currently using.
+  llvm::SpecificBumpPtrAllocator<EpilogueARCBlockState> BPA;
+
+  /// Current function we are analyzing.
+  SILFunction *F;
+
+  /// Current post-order we are using.
+  PostOrderFunctionInfo *PO;
+
+  /// Current alias analysis we are using.
+  AliasAnalysis *AA;
+
+  /// Current rc-identity we are using.
+  RCIdentityFunctionInfo *RCFI;
+
+  /// The epilogue retains or releases.
+  llvm::SmallVector<SILInstruction *, 1> EpilogueARCInsts; 
+
+  /// All the retain/release block state for all the basic blocks in the function. 
+  llvm::DenseMap<SILBasicBlock *, EpilogueARCBlockState *> EpilogueARCBlockStates;
+
+  /// The exit blocks of the function.
+  llvm::SmallPtrSet<SILBasicBlock *, 2> ExitBlocks;
+
+  /// Return true if this is a function exit block.
+  bool isExitBlock(SILBasicBlock *BB) {
+    return ExitBlocks.count(BB);
+  }
+
+  /// Return true if this is a retain instruction.
+  bool isRetainInstruction(SILInstruction *II) {
+    return isa<RetainValueInst>(II) || isa<StrongRetainInst>(II);
+  }
+
+  /// Return true if this is a release instruction.
+  bool isReleaseInstruction(SILInstruction *II) {
+    return isa<ReleaseValueInst>(II) || isa<StrongReleaseInst>(II);
+  }
+
+public:
+  /// Constructor.
+  EpilogueARCContext(EpilogueARCKind Kind, SILValue Arg, SILFunction *F,
+                     PostOrderFunctionInfo *PO, AliasAnalysis *AA,
+                     RCIdentityFunctionInfo *RCFI)
+    : Kind(Kind), Arg(Arg), F(F), PO(PO), AA(AA), RCFI(RCFI) {}
+
+  /// Run the data flow to find the epilogue retains or releases.
+  bool run() {
+    // Initialize the epilogue arc data flow context.
+    initializeDataflow();
+    // Converge the data flow.
+    convergeDataflow();
+    // Lastly, find the epilogue ARC instructions.
+    return computeEpilogueARC();
+  }
+
+  /// Reset the epilogue arc instructions. 
+  void resetEpilogueARCInsts() { EpilogueARCInsts.clear(); }
+  llvm::SmallVector<SILInstruction *, 1> getEpilogueARCInsts() {
+    return EpilogueARCInsts;
+  }
+
+  /// Initialize the data flow.
+  void initializeDataflow();
+
+  /// Keep iterating until the data flow is converged.
+  void convergeDataflow();
+
+  /// Find the epilogue ARC instructions.
+  bool computeEpilogueARC();
+
+  /// This instruction prevents looking further for epilogue retains on the
+  /// current path.
+  bool mayBlockEpilogueRetain(SILInstruction *II, SILValue Ptr) { 
+    // reference decrementing instruction prevents any retain to be identified as
+    // epilogue retains.
+    if (mayDecrementRefCount(II, Ptr, AA))
+      return true;
+    // Handle self-recursion. A self-recursion can be considered a +1 on the
+    // current argument.
```

Comment:
Because say you have func foo() -> @owned result and I am trying to find all the epilogue retains for the return result.  I treat func foo() -> @owned as the epilogue retain and FSO eventually turn the function into func foo() -> (not owned). Maybe i should implement this as a lambda which individual pass can pass to the mayBlocEpilogueRetain function